microprofile/fault-tolerance-smallrye/executor/src/main/java/org/wildfly/extension/microprofile/faulttolerance/MiniContextPropagation.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2020 JBoss by Red Hat.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.extension.microprofile.faulttolerance;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Miniature context propagation API, adapted from MicroProfile Context Propagation and SmallRye Context Propagation.
+ * @author Emmanuel Hugonnet (c) 2020 Red Hat, Inc.
```

Comment:
You stole my code! :-D

Really, I'd suggest to:
- make this class package-private
- add a big TODO that this should be replaced by Context Propagation proper when it's integrated in WildFly

Reply:
Certainly a note can be added, but it's not that straightforward – even though if Context Propagation is integrated in WildFly it doesn't mean that the users must provision it in theirs server while still anticipating that the tracing context propagation works - which is what we are doing in the PR.

Reply:
Yea, the thing is, what this PR does (which is the same thing I did in Thorntail) won't work with other things that also use context propagation. Thorntail is dead, but WildFly isn't, so it deserves better treatment. I'd suggest that if users provision Fault Tolerance and OpenTracing, but don't provision Context Propagation, then they should _NOT_ expect traces being propagated across FT `@Asynchronous` calls. Obviously, that becomes a breaking change if this PR is merged, so...

Again, I could do this hack in Thorntail, because it's dead, but you need to be more careful with WildFly.

Reply:
@Ladicek  You mean I took responsibility of this piece of code in WildFly ? ;)

Reply:
@Ladicek I disagree with your POV on `@Asynchronous` having tracer spans coming out of nowhere doesn't seem something that make sense. The fact that this issue arose is proof of that to me. If context propagation is provided we should rely on it insted of the current mechanism but otherwise I think we should still provide it.

Reply:
I agree with you, but the same argument applies to all cases where context propagation is / can be  used, such as RestClient + OpenTracing, @Asynchronous + @RequestScoped, etc. etc. etc. It doesn't make any sense to build a special case support for everything.

On the other hand, of course, Context Propagation is not in MicroProfile yet, so this is all a gray area.

Reply:
@rhusar WDYT here?

Reply:
I am on the fence here, spent a lot of time thinking about this, here is some summary:

* All of this is out of the spec – there is no mention in FT nor OpenTracing.
* We can't leverage Context Propagation for this as its not part of the platform yet – we will when it does make it to the platform.
* If we bend over backwards now to support this use case, going forward, we will have to deal with complexity once Context Propagation is available.
* Yet, it does make sense to propagate the context in this scenario, but if we are doing that, we should evaluate all cases across the board, rather than a single one.
* The original issue was reported by QE, not a real world user use case; so e.g. introducing this in XP2 is a behavior change which might not be expected.

I think it boils down to what users are expecting to happen and @Ladicek 's comment "I'd suggest that if users provision Fault Tolerance and OpenTracing, but don't provision Context Propagation, then they should NOT expect traces being propagated across FT @Asynchronous calls." suggests that we shouldn't move forward with this.